Fix error showing TaskRun steps that have not run

ncskier opened the pull request:
# Changes

Fixes #644
This fix shows a neutral "Not run" status for steps that follow a step with an
error. Previously, these steps showed a misleading green "Completed" status.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

ncskier commented:
/uncc @dibbles 
/uncc @vtereso 
/cc @a-roberts 
/cc @steveodonovan

ncskier commented:
Thanks for the feedback @AlanGreene! I'll update this PR to show the default step state instead of hiding the steps 👍 
/hold

AlanGreene commented on packages/components/src/components/Log/Log.js:
```diff
@@ -82,6 +82,11 @@ export class LogContainer extends Component {
           id: 'dashboard.pipelineRun.stepFailed',
           defaultMessage: 'Step failed'
         });
+      case 'Not run':
```
no need to display a trailer on the logs for this case

ncskier replied:
True, it's not really necessary. I'll remove it 👍

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -27,6 +27,7 @@ import {
 } from '@tektoncd/dashboard-utils';
 
 import { Log, RunHeader, StepDetails, TaskTree } from '..';
+import { updateUnexecutedSteps } from '../../../../../src/utils';
```
code in packages/* cannot consume code directly from the core dashboard, it must be available in a published npm package (in this case it may make sense to move it to packages/utils) otherwise it will not work for 3rd parties consuming the @tektoncd/dashboard-components package.

ncskier replied:
@AlanGreene So, what would be the best way for me to make this function available? It's used in the following files:
- `src/containers/TaskRun/TaskRun.js`
- `packages/components/src/components/Task/Task.js`
- `packages/components/src/components/PipelineRun/PipelineRun.js`

If I move the function declaration to `packages/utils`, then can I import it from `src/containers/TaskRun/TaskRun.js`?

AlanGreene replied:
yes that'll work 👍 
you'll just need to import it from `@tektoncd/dashboard-utils` rather than by local path

AlanGreene commented on packages/components/src/components/Task/Task.js:
```diff
@@ -89,24 +91,28 @@ class Task extends Component {
+                    key={id}
+                    onSelect={this.handleStepSelected}
+                    reason={stepReason}
+                    // reason={index > errorIndex ? '' : stepReason}
```
remove

AlanGreene commented on src/utils/index.js:
```diff
@@ -74,3 +74,36 @@ export function getGitValues(url) {
+
+// Update the status of steps that follow a step with an error
+export function updateUnexecutedSteps(steps) {
+  if (typeof steps === 'undefined') {
```
is there any change `steps` could be null? if so might be safer just to check `if (!steps) {`

ncskier replied:
Ok, thanks. I changed it. 

How about for these lines (https://github.com/tektoncd/dashboard/pull/653/files#diff-44384c8c75c7b5aab1477fe8f0736133R162-R163):
```
typeof step.stepStatus !== 'undefined' &&
typeof step.stepStatus.terminated !== 'undefined'
```
@AlanGreene Is it safer to do `if (!step.stepStatus && step.stepStatus.terminated)...` ?

AlanGreene replied:
`if (step.stepStatus && step.stepStatus.terminated) {`

yes, easier to read / reason about too (IMO)

ncskier replied:
Ok, thanks! 👍

AlanGreene commented on src/utils/index.js:
```diff
@@ -74,3 +74,36 @@ export function getGitValues(url) {
+    return steps;
+  }
+  let errorIndex = steps.length - 1;
+  return steps.map((step, index) => {
```
I think it's worth raising https://github.com/tektoncd/pipeline/issues/1439 in the next working group meeting at least to ask for feedback on the issue.

It also looks like the behaviour when cancelling a PipelineRun is different, I see steps marked as Waiting instead of Completed.

Having this workaround in the dashboard for now resolves the immediate user-facing issue for users of the dashboard, but users of kubectl or other tools still have the same misleading information presented (i.e. a step that hasn't executed is marked as Completed), and each tool needs to handle this themselves.

AlanGreene commented:
@ncskier can you also update the PR title + description to reflect the change in approach taken here? Thanks

ncskier commented:
Thanks for the feedback @AlanGreene! I've updated the PR title & description, and I think I've addressed all of your concerns. Let me know if there's anything else that I can do 🙂

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -27,6 +27,7 @@ import {
 } from '@tektoncd/dashboard-utils';
 
 import { Log, RunHeader, StepDetails, TaskTree } from '..';
+import { updateUnexecutedSteps } from '../../../../utils';
```
import from '@tektoncd/dashboard-utils'

AlanGreene commented on packages/components/src/components/Task/Task.js:
```diff
@@ -17,6 +17,8 @@ import ChevronRight from '@carbon/icons-react/lib/chevron--right/16';
 import CloseFilled from '@carbon/icons-react/lib/close--filled/16';
 import { Spinner, Step } from '@tektoncd/dashboard-components';
 
+import { updateUnexecutedSteps } from '../../../../utils';
```
import from '@tektoncd/dashboard-utils'

ncskier commented:
Ah, thanks @AlanGreene! Just fixed the imports 😁
